### PR TITLE
[BEAM-2253] Use resource transformer plugin to use versions defined in root pom.xml within maven archetype poms

### PR DESCRIPTION
**Pull request by lukecwik**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`.
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @kennknowles @tgroh

File: sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml [hunk 1]
```diff
@@ -50,7 +49,7 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
-        <version>3.5.1</version>
+        <version>@maven-compiler-plugin.version@</version>
```
Comment (kennknowles) [inline sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml]: This is fine, for sure. But did you consider adding them as properties up top? The silly looking pattern of `<surefire-plugin.version>@surefire-plugin.version</surefire-plugin.version>`. The layer of indirection seems to be a pretty useful practice for keeping versions close to each other and easy to read.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml [hunk 2]
```diff
@@ -60,7 +59,7 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-surefire-plugin</artifactId>
-        <version>${surefire-plugin.version}</version>
+        <version>@surefire-plugin.version@</version>
         <configuration>
           <parallel>all</parallel>
           <threadCount>4</threadCount>
```
Comment (dhalperi) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: missed L72 below.

File: pom.xml [hunk 3]
```diff
@@ -143,8 +143,11 @@
     <groovy-maven-plugin.version>2.0</groovy-maven-plugin.version>
     <surefire-plugin.version>2.20</surefire-plugin.version>
     <failsafe-plugin.version>2.20</failsafe-plugin.version>
+    <maven-compiler-plugin.version>3.6.1</maven-compiler-plugin.version>
```
Comment (dhalperi) [inline pom.xml]: would prefer consistent naming. We seem to have both forms for `org.apache.maven.plugins:maven-x-plugin`:

* ${maven-x-plugin}
* ${x-plugin}

Comment (dhalperi) [inline pom.xml]: (I guess this PR does not make things worse)

File: sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml (hunk 1 above)
Comment (dhalperi) [inline sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml]: +1

File: pom.xml (hunk 3 above)
Comment (kennknowles) [inline pom.xml]: Aside from the coupled `google-clients.version` I would favor naming things after their `artifactId` to make things completely obvious and automatic.

File: sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml (hunk 1 above)
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/pom.xml]: I was thinking it was easier for a reader of the code to understand that the @yyy.version@ is coming from outside without needing to read the properties section. Having the properties section allows for maven archetype users to have a single property location.

Will update and repost.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml (hunk 2 above)
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: Done

File: pom.xml (hunk 3 above)
Comment (lukecwik) [inline pom.xml]: Sounds like a good follow up. Filed https://issues.apache.org/jira/browse/BEAM-2327 to track.

**Pull request closed without merging**